crates/deltalake-core/src/operations/transaction/protocol.rs

```diff
@@ -0,0 +1,404 @@
+use std::collections::HashSet;
+
+use lazy_static::lazy_static;
+use once_cell::sync::Lazy;
+
+use super::TransactionError;
+use crate::kernel::{Action, ReaderFeatures, WriterFeatures};
+use crate::table::state::DeltaTableState;
+
+lazy_static! {
```

Comment:
What is this used for? The table reader and writers have `Into<usize>` impls so you should just be able to cast them to find their associated reader / writer version.

Reply:
Main intent is to have a uniform way of handling compatibility for tables on 3/7 and the ones below.

Reply:
Sorry I think I might not understand the check here. Are you saying these are for the chain of required features? Like check constraints for example requires append only and invariants? So maybe it's the name that throws me off if that's the case.

Reply:
as usual, it depends 😆 - tables features as a concept are only available at reader/writer version 3/7. AT that version, clients can validate that they implement all features the table requires as stated in the `Protocol` action. The respective fields must be available at these versions. For any reader/writer version below that, clients must assume that they need to implement all features up to the respective version.

Reply:
Ahhh I see, so this only matters when we are not dealing with actual table features and meant to enforce the "all or nothing" style of protocol versioning prior?

Reply:
Exactly.